Move from a "github config" to a "munge object"

eparis opened the pull request:
- Label removals in one munger will be 'seen' in the next
- Locally cache github results and use conditional gets
- Significant improvements to submit-queue web interface (won't crash your browser after a while)
- More debug and visibility information for users concerning what the bot is up to.

eparis committed:
mungegithub: create a struct which encapsulates github information

This removes the need to import the go-github api into most mungers.

eparis committed:
Make github take mungeObjects instead of go-github objects

eparis committed:
Remove 'issues' and rename 'pulls' to 'mungers'

eparis committed:
Remove unused 'labels' from the ForEachPR loop

It was build to be able to only process issues with the given labels.
But we never used it. Get rid of it.

eparis committed:
Simplify more code as we remove the 'issues' v 'pulls' differentiation

eparis committed:
Add a 'config' to the munge object

Really we want to work on a munge object, not a 'config'.  So add the
config to the munge object.

eparis committed:
Convert to using the config in the MungeObject

eparis committed:
Use helpers to get at parts or MungeObject

This start down the path of lazily getting information

eparis committed:
Make most of MungeObject private

Only let mungers mess with the Issue. Helpers for everything else.
Another step on lazy initialization.

eparis committed:
Only get PR, Events, and Commits as needed

eparis committed:
Minor code order change to exit having done less work

eparis committed:
submit_queue: use obj.issue not PR

eparis committed:
Fix perfdash after great munger move

eparis commented:
@lavalamp ask to do reviews and you shall receive   :)

eparis committed:
Return support for non-PR mungers

Do this by testing if it is a PR inside each munger

eparis committed:
Rework test functions

Create a central server which can server out issues, prs, status,
commits, events, etc. Also place the helper functions to create those
objects in a single place.

While figuring out the minimal number of things needed in some objects
we hit a number of panics in the code. These are things that are likely
impossible when dealing with github, but add some more nil checks, since
my bad reimplementation of github hit them.

eparis committed:
Store analytics for display by mungers

Display bot github api usage on the web page

eparis committed:
Stop needlessly passing github config to EachLoop

eparis committed:
Display submit queue history (up to 128 entries)

We don't show an entry for everything, just those that at least got to
the github e2e re-test

eparis committed:
Break users into their own API

eparis committed:
Break stats into their own api

(and refresh much less often!)

eparis committed:
Break the current github e2e queue into its own api

eparis committed:
Move google internal CI to seperate api

eparis committed:
Rename "api" to "prs" since it is just prs now

Also, update much much less often, as that isn't even the most
interesting information any more.

eparis committed:
Remove labels from object when API call made

This means that if one mungers removes a label, other's won't see it!

eparis committed:
Try to use httpcache again

eparis committed:
Store the github api limits when we reset counters

eparis committed:
Implement our own github API limit controller

It leaves 500 api calls unused. If we run out of calls it will wait
until the tokens reset. Given the local http cache, and this better
tracking (as each github api call no longer necessarily counts against
the api, so a time based rate limiter pegged to the number of github api
calls per second doesn't really make sense).

Numbers appear to indicate that we can now run every 10 minutes!

eparis committed:
Print how often we got an answer from the cache

eparis committed:
Better object tracking in angular

eparis committed:
Search box on history page

eparis committed:
Do filtering with an angular filter rather than ng-show

This means we don't mutate objects. So even though I added track-by to
PRs angular still couldn't really track history at all. Now since the
object is static it should be able to...

eparis committed:
Harmonize logging in submit queue: make all status updates V(4)

eparis committed:
Enable the httpcache by default and go faster!

With the cache we can run every 10 minutes!

eparis committed:
Use the kubernetes WordSepNormalizeFunc for consistent flag names

And as a side effect, it returns kubernetes/pkg/util/ in the import which
means that we automatically set glog to log to stderr.

eparis commented:
@lavalamp @brendandburns huge improvements, including being able to run the bot every 10 minutes instead of every 30 and the web site won't crash your browser after a while any more.

Can we get a merge/build/run and if it fails, just roll back. The bot won't do anything new/crazier than it did, but we'll have a lot more idea what it's up to...

eparis commented:
http://plnkr.co/edit/M9xY3oM8e9VChnPXJsFt?p=preview is a pretty up2date example of the UI features.

eparis commented on mungegithub/github/github.go:
```diff
@@ -38,28 +38,90 @@ import (
+// a super fresh copy. But since all of the other calls are only going to be made
+// once per poll loop the 60 second github freshness doesn't matter. So I can't
+// think of a reason not to just keep this simple and always set max-age=0 on
+// every request.
```
discussion why i believe we can start using gregjones/httpcache again (to get the huge speedup)

lavalamp commented:
Wow I'm not sure how this evaded my filter the first time around, but I didn't see it until just now. Sorry!

@brendandburns can you take this or delegate? I'm going to be OOO tomorrow.

eparis committed:
Do not keep trying to re-open and open PR

The logic was wrong. We were supposed to exit early on success, instead
we were exitting early on failure!

eparis committed:
Remove useless complaint in blunderbuss

We do not use the SHA or Author.Login so why would we care if they are
filled in or not? (In fact SHA == nil would never get here, it is
checked in the github helper code). Stop spamming the logs.

eparis committed:
Log every time we change PR

at glog.Info().  And don't log those things (at and verbose level)
inside the mungers

eparis committed:
Change automated tagging schede to be date+git

eparis committed:
Quiet down when we have to wait for mergeable data

This isn't something actionable

eparis committed:
Rename pod.yaml to rc.yaml

(But no file changes, just for nice history)

eparis committed:
Update from a pod to an rc

So we can resize to 0 and leave it for rollback

eparis committed:
Do not update status if the PR is already in the queue (or running)

eparis committed:
Fix user list to properly show 'pull' instead of 'explicit'

And update the user lists

eparis committed:
Makefile update to run a local dryrun container

eparis committed:
Fix clicking on user on users page

eparis committed:
Silence some possible CI problems

It complains that it can't pull contributors, so shut that up.
We also had a race. If the PR was brought off the queue before we
checked its status it would 'fail'. We can't really check (easily) that
the PR is queue message was properly set since out test code is too
fast. So don't test for it. If we forget to grab that status, no big
deal.
